### ISSUE TITLE
[$250] Track expense - App crashes on report details RHP after submitting track expense to workspace

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#53547]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.71-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:** Exp https://expensify.testrail.io/index.php?/tests/view/5299012
**Email or phone of affected tester (no customers):** applausetester+kh1311020@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
Precondition:
- Account has a workspace.
- Workspace has no unsettled expenses.

1. Go to staging.new.expensify.com
2. Go offline.
3. Go to self DM.
4. Track a manual expense.
5. Click Categorize it from the actionable whisper.
6. Select a category, enter merchant and submit the expense.
7. Go to transaction thread in the workspace chat.
8. Click on the report header.
9. Go online while staying on the report details RHP.

## Expected Result:
App will not crash.

## Actual Result:
App crashes.
Tester's devices where the issue occurs:
 macOS Ventura 13.4 - Chrome
 Samsung Galaxy Z Fold4 - Android 14,
 Apple iPhone 15 Pro Max - iOS 18.1 Beta.

## Workaround:
Unknown

## Platforms:
- [x] Android: Standalone
- [x] Android: HybridApp
- [x] Android: mWeb Chrome
- [x] iOS: Standalone
- [x] iOS: HybridApp
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/5fd0d11d-a540-4be4-a0b3-68e7c2df5d70

[Bug6684140_1733307605993!staging.new.expensify.com-1733307405873.txt](https://github.com/user-attachments/files/18007638/Bug6684140_1733307605993.staging.new.expensify.com-1733307405873.txt)

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021864495667152345213</li>
        <li>Upwork Job ID: 1864495667152345213</li>
        <li>Last Price Increase: 2024-12-05</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @hoangzinh</details>

**Comment by saifelance** [comment]
### Please re-state the problem that we are trying to solve in this issue.
The app crashes on the report details Right-Hand Panel (RHP) after submitting a tracked expense to a workspace offline and going online, disrupting user access to the report.

### What is the root cause of that problem?

1. **Offline-to-Online Transition**: State changes during this transition are not handled, causing invalid or incomplete data rendering.
2. **Undefined Data**: Missing or incomplete report or parentAction objects cause crashes when accessed.

### What changes do you think we should make in order to solve the problem?
`src/components/ParentNavigationSubtitle.tsx`

```
function ParentNavigationSubtitle({
    parentNavigationSubtitleData,
    parentReportActionID,
    parentReportID = '',
    pressableStyles,
}: ParentNavigationSubtitleProps) {
    const styles = useThemeStyles();
    const { workspaceName, reportName } = parentNavigationSubtitleData;
    const { isOffline } = useNetwork();
    const { translate } = useLocalize();
    const [report] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT}${parentReportID}`);
    const canUserPerformWriteAction = ReportUtils.canUserPerformWriteAction(report);

    // Guard clause for missing report name
    if (!reportName) {
        return;
    }

    // Validate report and parentAction
    if (!report || !reportName) {
        console.error("Missing report data", { parentReportID });
        return;
    }

    const parentAction = ReportActionsUtils.getReportAction(parentReportID, parentReportActionID ?? '-1');
    if (!parentAction) {
        console.error("Missing parent action data", { parentReportID, parentReportActionID });
        return;
    }

    // Prevent navigation if offline
    if (isOffline) {
        console.warn("Cannot navigate while offline");
        return;
    }

    return (
        <PressableWithoutFeedback
            onPress={() => {
                const isVisibleAction = ReportActionsUtils.shouldReportActionBeVisible(
                    parentAction, parentAction?.reportActionID ?? '-1', canUserPerformWriteAction
                );

                // Pop the thread report screen before navigating to the chat report.
                Navigation.goBack(ROUTES.REPORT_WITH_ID.getRoute(parentReportID));

                if (isVisibleAction) {
                    // Pop the chat report screen before navigating to the linked report action.
                    Navigation.goBack(ROUTES.REPORT_WITH_ID.getRoute(parentReportID, parentReportActionID), true);
                } else {
                    console.warn("Report action is not visible");
                }
            }}
            accessibilityLabel={translate('threads.parentNavigationSummary', { reportName, workspaceName })}
            role={CONST.ROLE.LINK}
            style={pressableStyles}
        >
            <Text
                style={[styles.optionAlternateText, styles.textLabelSupporting]}
                numberOfLines={1}
            >
                {!!reportName && (
                    <>
                        <Text style={[styles.optionAlternateText, styles.textLabelSupporting]}>{`${translate('threads.from')} `}</Text>
                        <Text style={[styles.optionAlternateText, styles.textLabelSupporting, styles.link]}>{reportName}</Text>
                    </>
                )}
                {!!workspaceName && <Text style={[styles.optionAlternateText, styles.textLabelSupporting]}>{` ${translate('threads.in')} ${workspaceName}`}</Text>}
            </Text>
        </PressableWithoutFeedback>
    );
}
```

- **Null/Undefined checks** added for `report`, `parentAction`, `parentReportID`, and `parentReportActionID`.
- **Offline state checks** to prevent navigation when offline.
- **Early returns** if data is invalid, preventing further execution and potential crashes.

### What alternative solutions did you explore? (Optional)

1. **Defer Navigation**: Wait until all data is loaded before rendering or navigating.
2. **Server-Side Validation**: Use server-side checks to mitigate client-side data issues.
3. **Error Boundary**: Implement an error boundary to catch crashes and prevent app failure.

**Comment by adelekennedy** [comment]
reproduced

**Comment by saifelance** [comment]
> Job added to Upwork: https://www.upwork.com/jobs/~021864495667152345213

Send invite please also previously are pending?

**Comment by mkzie2** [comment]
@hoangzinh We have a problem from the backend, after we go offline and online, the `CategorizeTrackedExpense` API returns `parentReportID` and `parentReportActionID` of the expense report is `null`. This causes `parentReportID` in `ParentNavigationSubtitle` to change from an existing ID to an empty string.

**Comment by saifelance** [comment]
@hoangzinh The backend returns correct IDs most of the time, but during offline-to-online transitions, the state handling is inconsistent, leading to null values for these fields.

**Comment by saifelance** [comment]
**1. Null/Undefined Check for Report Data**

```
// Check if report is valid before proceeding
if (!report || !reportName) {
    console.error("Missing report data", { parentReportID });
    return;
}

// Check if parentAction is valid
const parentAction = ReportActionsUtils.getReportAction(parentReportID, parentReportActionID ?? '-1');
if (!parentAction) {
    console.error("Missing parent action data", { parentReportID, parentReportActionID });
    return;
}
```

**2. Offline/Online Network State Handling**

```
// Prevent navigation if offline
if (isOffline) {
    console.warn("Cannot navigate while offline");
    return;
}

// Proceed with navigation if online
const isVisibleAction = ReportActionsUtils.shouldReportActionBeVisible(
    parentAction, parentAction?.reportActionID ?? '-1', canUserPerformWriteAction
);
if (isVisibleAction) {
    Navigation.goBack(ROUTES.REPORT_WITH_ID.getRoute(parentReportID, parentReportActionID), true);
} else {
    console.warn("Report action is not visible");
}
```

**3. Handle Missing Data for parentReportID and parentReportActionID**

```
// Validate parentReportID and parentReportActionID
if (!parentReportID || !parentReportActionID) {
    console.error("Missing required IDs: parentReportID or parentReportActionID");
    return;
}
```

**4. General Safe Data Access**

```
// General check for undefined data
if (!parentNavigationSubtitleData || !parentNavigationSubtitleData.reportName) {
    console.error("Missing subtitle data");
    return;
}
```
### Summary of Changes:

- Added null/undefined checks for `report`, `parentAction`, `parentReportID`, and `parentReportActionID`.
- Added offline checks to block navigation when offline.
- Implemented early returns for invalid data to avoid crashes.

@hoangzinh Could you please check if this helps?

**Comment by mkzie2** [comment]
@hoangzinh You're right, the problem comes from `ReconnectApp` API. The last update has `parentReportID` is `null`. I updated my [proposal](https://github.com/Expensify/App/issues/53547#issuecomment-2517186439).

https://github.com/user-attachments/assets/98ba28b3-df7e-4861-b852-c5065bfda85a

**Comment by saifelance** [comment]
The issue stems from the `ReconnectApp` API, where the last update intermittently has `parentReportID` set to null. This causes the `ParentNavigationSubtitle` to lose its previous valid `parentReportID`, leading to inconsistent data display in the frontend.

**Comment by saifelance** [comment]
@hoangzinh @adelekennedy 

@mkzie2 
**===>The last update has parentReportID is null. I updated my https://github.com/Expensify/App/issues/53547#issuecomment-2517186439.**

> No its required backend fixing only

The frontend relies on `parentReportID` and `parentReportActionID`, but null values from the backend cause navigation issues. Adding a fallback mechanism in the frontend can stabilize behavior temporarily. However, a backend fix is essential for a long-term solution.

**Comment by mkzie2** [comment]
> but it causes an issue with navigation back to workspace being hidden.

@hoangzinh I agree this is a backend bug. But we still need a frontend fix as my solution to prevent the app crashes in some cases with the same problem. For example, when we change the value in the debug mode:


https://github.com/user-attachments/assets/080abb25-cd51-4939-9909-90dddd8552c3

**Comment by saifelance** [comment]
> @saifelance I think we just need to fallback the `parentReportID` to -1 as @mkzie2 suggested in his proposal is enough if we want to avoid app crash. Btw, just wait internal engineer thoughts on it.

@hoangzinh If you only need to apply a fallback for `parentReportID` (and not for `parentReportActionID`),

@adelekennedy  suggestion to fallback to '`-1`' when `parentReportID` is null or empty is a reasonable temporary fix. The crash happens because the `useOnyx` hook expects a valid `parentReportID`, and null or an empty string causes errors.

The proposed solution (fallback to `'-1')` is a good temporary fix to prevent app crashes. However, the root cause (backend behavior after reconnecting) should be addressed for a long-term fix.

**Comment by mkzie2** [comment]
@hoangzinh I think we can add a C+ reviewed comment to get the eye from internal team.

**Comment by maddylewis** [comment]
@adelekennedy - when you have a sec, could you write the responses to these prompts? and then ill add this to the HOT PICKS section of the Retain update today 👍

https://stackoverflowteams.com/c/expensify/questions/20462

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by adelekennedy** [comment]
I can no longer reproduce this - I think we are safe to close!

**Comment by vincdargento** [comment]
@adelekennedy Tester is still able to reproduce the issue. Re-opening in case you want to take a look

https://github.com/user-attachments/assets/9962d45f-dc78-4e3c-b99e-5089f142b8fe

**Comment by mkzie2** [comment]
@hoangzinh Since the backend change could be low-prioritized and the bug can reproduce by another way [here](https://github.com/Expensify/App/issues/53547#issuecomment-2527494357), we can move forward with the frontend fix in my [solution](https://github.com/Expensify/App/issues/53547#issuecomment-2517186439). What do you think?

**Comment by maddylewis** [comment]
@mkzie2 - to confirm, is this issue (5347) a duplicate of this known issue -- https://github.com/Expensify/App/issues/54251?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by adelekennedy** [comment]
I'm not able to reproduce on IOS web or browserstack - @hoangzinh do you mind giving it a try? If this was fixed by another issue you're certainly still due compensation here!

**Comment by adelekennedy** [comment]
great! I think we can close this then - @hoangzinh I think 50% is due here as you've been really diligent moving this forward even without a PR being raised.

Contributor+: @hoangzinh owed $125.00 via NewDot

**Comment by garrettmknight** [comment]
$125 approved fro @hoangzinh
